### PR TITLE
[tmf] add method to read entry from ChildTableTlv

### PR DESCRIPTION
**Pull request by huamenggg**
This commit add a method to read entry from ChildTableTlv. The GetEntry() cannot be used because the current read ChildTableTlv operation doesn't read child table entries.

### CODE REVIEW

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 1]
```diff
@@ -1348,6 +1348,24 @@ class ChildTableTlv : public NetworkDiagnosticTlv
         return *reinterpret_cast<ChildTableEntry *>(GetValue() + (aIndex * sizeof(ChildTableEntry)));
     }
 
+    /**
+     * This method reads the Child Table entry at @p aIndex.
+     *
+     * @param[in]  aIndex  The index into the Child Table list.
+     *
+     * @returns  if read successfully.
```
Comment (bukepo) [inline src/core/thread/network_diagnostic_tlvs.hpp]: better to document possible errors. for example:
```
@retval OT_ERROR_NOT_FOUND No such entry is found.
@retval OT_ERROR_NONE Successfully read the entry.
```

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 2]
```diff
@@ -1348,6 +1348,24 @@ class ChildTableTlv : public NetworkDiagnosticTlv
...
+     */
+    otError ReadEntry(ChildTableEntry &aEntry, Message &aMessage, uint16_t aOffset, uint8_t aIndex)
+    {
+        otError error = OT_ERROR_PARSE;
```
Comment (bukepo) [inline src/core/thread/network_diagnostic_tlvs.hpp]: use `OT_ERROR_NOT_FOUND` instead?

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 3]
```diff
@@ -1348,6 +1348,24 @@ class ChildTableTlv : public NetworkDiagnosticTlv
         return *reinterpret_cast<ChildTableEntry *>(GetValue() + (aIndex * sizeof(ChildTableEntry)));
     }
 
+    /**
+     * This method reads the Child Table entry at @p aIndex.
+     *
+     * @param[in]  aIndex  The index into the Child Table list.
```
Comment (librasungirl) [inline src/core/thread/network_diagnostic_tlvs.hpp]: add doxygen for the other parameters?

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 4]
```diff
@@ -1348,6 +1348,24 @@ class ChildTableTlv : public NetworkDiagnosticTlv
...
+     *
+     * @returns  if read successfully.
+     */
+    otError ReadEntry(ChildTableEntry &aEntry, Message &aMessage, uint16_t aOffset, uint8_t aIndex)
```
Comment (librasungirl) [inline src/core/thread/network_diagnostic_tlvs.hpp]: Could existing `GetEntry()` in Line 1346 serve the same purpose?

File: src/posix/platform/udp.cpp [hunk 5]
```diff
@@ -71,7 +71,8 @@ static int FdFromHandle(void *aHandle)
 
 static bool IsLinkLocal(const struct in6_addr &aAddress)
 {
-    return aAddress.s6_addr[0] == 0xfe && aAddress.s6_addr[1] == 0x80;
+    return (aAddress.s6_addr[0] == 0xfe && aAddress.s6_addr[1] == 0x80) ||
+           (aAddress.s6_addr[0] == 0xff && aAddress.s6_addr[1] == 0x02);
```
Comment (librasungirl) [inline src/posix/platform/udp.cpp]: In OT's existing implementation, `IsLinkLocal` is used to check whether a unicast address is link local and there would be a separate `IsLinkLocalMulticast` to check whether a multicast address is link local. 

https://github.com/openthread/openthread/blob/master/src/core/net/ip6_address.cpp#L65
https://github.com/openthread/openthread/blob/master/src/core/net/ip6_address.cpp#L75

Better to keep consistent for such interfaces?

File: src/core/thread/network_diagnostic_tlvs.hpp (hunk 4 above)
Comment (bukepo) [reply src/core/thread/network_diagnostic_tlvs.hpp]: discussed offline, `GetEntry()` cannot serve it.

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 6]
```diff
@@ -1348,6 +1348,27 @@ class ChildTableTlv : public NetworkDiagnosticTlv
         return *reinterpret_cast<ChildTableEntry *>(GetValue() + (aIndex * sizeof(ChildTableEntry)));
     }
 
+    /**
+     * This method reads the Child Table entry at @p aIndex.
+     *
```
Comment (librasungirl) [inline src/core/thread/network_diagnostic_tlvs.hpp]: `aEntry` is missed.

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 7]
```diff
@@ -1348,6 +1348,27 @@ class ChildTableTlv : public NetworkDiagnosticTlv
...
+     * @param[in]   aMessage    A reference to the message.
+     * @param[out]  aOffset     The offset of the ChildTableTLV in aMessage.
```
Comment (librasungirl) [inline src/core/thread/network_diagnostic_tlvs.hpp]: `aOffset` should be `param[in]?`

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 8]
```diff
@@ -1348,6 +1348,28 @@ class ChildTableTlv : public NetworkDiagnosticTlv
         return *reinterpret_cast<ChildTableEntry *>(GetValue() + (aIndex * sizeof(ChildTableEntry)));
     }
 
+    /**
+     * This method reads the Child Table entry at @p aIndex.
+     *
+     * @param[in]   aEntry      A reference to a ChidTableEntry.
```
Comment (bukepo) [inline src/core/thread/network_diagnostic_tlvs.hpp]: => `@param[out]`
typo: ChildTableEntry.

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 9]
```diff
@@ -1348,6 +1348,28 @@ class ChildTableTlv : public NetworkDiagnosticTlv
...
+     * @param[in]   aOffset     The offset of the ChildTableTLV in aMessage.
+     * @param[in]   aIndex      The index into the Child Table list.
+     *
+     * @retval  OT_ERROR_NOT_FOUND  No such entry is found.
+     * @retval  OT_ERROR_NONE       Successfully read the entry.
+     */
+    otError ReadEntry(ChildTableEntry &aEntry, Message &aMessage, uint16_t aOffset, uint8_t aIndex)
+    {
+        otError error = OT_ERROR_NOT_FOUND;
+        if (aMessage.Read(aOffset + sizeof(ChildTableTlv) + (aIndex * sizeof(ChildTableEntry)), sizeof(ChildTableEntry),
```
Comment (bukepo) [inline src/core/thread/network_diagnostic_tlvs.hpp]: Verify `aIndex < GetNumEntries()`?

File: src/posix/platform/udp.cpp [hunk 10]
```diff
@@ -71,7 +71,7 @@ static int FdFromHandle(void *aHandle)
 
 static bool IsLinkLocal(const struct in6_addr &aAddress)
 {
-    return aAddress.s6_addr[0] == 0xfe && aAddress.s6_addr[1] == 0x80;
+    return (aAddress.s6_addr[0] == 0xfe && aAddress.s6_addr[1] == 0x80);
```
Comment (bukepo) [inline src/posix/platform/udp.cpp]: revert the change?

File: src/core/thread/network_diagnostic_tlvs.hpp (hunk 8 above)
Comment (bukepo) [reply src/core/thread/network_diagnostic_tlvs.hpp]: @huamenggg typo is not corrected yet.

File: src/core/thread/network_diagnostic_tlvs.hpp [hunk 11]
```diff
@@ -1348,6 +1348,33 @@ class ChildTableTlv : public NetworkDiagnosticTlv
...
+     *
+     * @param[out]  aEntry      A reference to a ChildTableEntry.
+     * @param[in]   aMessage    A reference to the message.
+     * @param[in]   aOffset     The offset of the ChildTableTLV in aMessage.
+     * @param[in]   aIndex      The index into the Child Table list.
+     *
+     * @retval  OT_ERROR_NOT_FOUND  No such entry is found.
+     * @retval  OT_ERROR_NONE       Successfully read the entry.
+     */
+    otError ReadEntry(ChildTableEntry &aEntry, Message &aMessage, uint16_t aOffset, uint8_t aIndex) const
```
Comment (jwhui) [inline src/core/thread/network_diagnostic_tlvs.hpp]: This method seems more complex than a basic get/set. Thoughts on moving this to a cpp file?
